test/integration/scheduler/volume_binding_test.go

```diff
@@ -361,23 +474,110 @@ func TestPVAffinityConflict(t *testing.T) {
 	}
 }
 
-func setupCluster(t *testing.T, nsName string, numberOfNodes int) *testConfig {
-	// Enable feature gates
-	utilfeature.DefaultFeatureGate.Set("VolumeScheduling=true,PersistentLocalVolumes=true")
+// TestRescheduleProvisioning tests:
+// 1. If scheduler will trigger provisioning by adding "selected-node" annotation
+// 2. If PV controller will reschedule provisioning that failed to provision
+// by removing the annotation.
+// To clearly verify the annotation is added/removed, we need to start/stop
+// scheduler and PV controller individually.
+func TestRescheduleProvisioning(t *testing.T) {
+	utilfeature.DefaultFeatureGate.Set("VolumeScheduling=true,DynamicProvisioningScheduling=true,PersistentLocalVolumes=true")
+	schedulerStopCh := make(chan struct{})
+	schedulerStopChOpen := true
+	controllerStopCh := make(chan struct{})
 
-	controllerCh := make(chan struct{})
-
-	context := initTestScheduler(t, initTestMaster(t, nsName, nil), controllerCh, false, nil)
+	context := initTestScheduler(t, initTestMaster(t, "reschedule-volume-provision", nil), schedulerStopCh, false, nil)
 
 	clientset := context.clientSet
 	ns := context.ns.Name
+
+	defer func() {
+		if schedulerStopChOpen {
+			close(schedulerStopCh)
+		}
+		cleanUpTestObjects(clientset, ns, nil)
+		context.clientSet.CoreV1().Nodes().DeleteCollection(nil, metav1.ListOptions{})
+		context.closeFn()
+		close(controllerStopCh)
+		utilfeature.DefaultFeatureGate.Set("VolumeScheduling=false,LocalPersistentVolumes=false,DynamicProvisioningScheduling=false")
+	}()
+
+	initSharedObjects(t, context, 1)
+
+	ctrl, err := createPVController(context)
+	if err != nil {
+		t.Fatalf("Failed to create PV controller: %v", err)
+	}
+
+	// Create a pvc that is expected to fail in provisioning.
+	pvcName := "pvc-fail-to-provision"
+	pvc := makePVC(pvcName, ns, &classFailProvision, "")
+	pvc, err = clientset.CoreV1().PersistentVolumeClaims(ns).Create(pvc)
+	if err != nil {
+		t.Fatalf("Failed to create PersistentVolumeClaim %q: %v", pvc.Name, err)
+	}
+
+	// Create a pod to make use of the pvc.
+	pod := makePod("pod-to-use-pvc", ns, []string{pvcName})
+	if _, err := clientset.CoreV1().Pods(ns).Create(pod); err != nil {
+		t.Fatalf("Failed to create Pod %q: %v", pod.Name, err)
+	}
+
+	// PVC can not be provisioned, pod should be unschedulable.
+	if err := waitForPodUnschedulable(clientset, pod); err != nil {
+		t.Errorf("Pod %q was not unschedulable: %v", pod.Name, err)
+	}
+
+	// Get the pvc after update.
+	pvc, err = clientset.CoreV1().PersistentVolumeClaims(ns).Get(pvcName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get PersistentVolumeClaim %q: %v", pvcName, err)
+	}
+
+	// Validate the provisioning annotation should have been added by scheduler.
+	if err := validateProvisionAnn(pvc, true); err != nil {
+		t.Errorf("Error in provision: %v", err)
+	}
+
+	// Stop scheduler first to avoid adding the annotation back.
+	close(schedulerStopCh)
+	schedulerStopChOpen = false
```

Comment:
good point

Reply:
Having said that, I haven't found proper way to ensure it's fully stopped. Do we have similar practices before?

Reply:
Hmmm I don't see a great way to check if everything is stopped without adding "running" flags into every go routine that gets started.

Maybe this behavior can be tested as a unit test instead of an integration test?

Reply:
@lichuqiang are you able to find a solution for this?

Reply:
Not yet, I'll investigate if unit test can help cover the reschedule scenario